Optimize FP8 MLA KV cache writes with Triton kernel

harvenstar opened the pull request:
# Updated Summary

  Optimize FP8 MLA KV cache writes in NSA decode by:
  - skipping `torch.cat([k_nope, k_rope])` in the FP8 path, and
  - reusing the existing `set_mla_kv_buffer_triton` two-tensor write kernel.

  **No new Triton write kernel is added in the final version.**

  ## Motivation

  In DeepSeek-V3.2 decode, each step writes FP8 MLA KV cache per layer. The baseline path incurs noticeable CPU-side overhead when using PyTorch fancy indexing (`index_put_`) for small writes.

  ## Changes

  1. **Add `quantize_k_cache_separate(k_nope, k_rope)`** in `python/sglang/srt/layers/attention/nsa/quant_k_cache.py`.
     - Reuses existing fast quantization kernel (`_quantize_k_cache_fast`) without requiring an upstream concat.
     - Returns `(nope_part, rope_part)` as uint8 byte views, matching the MLA KV buffer layout.

  2. **Update FP8 NSA KV write path** in `python/sglang/srt/mem_cache/memory_pool.py`:
     - Remove upstream concat.
     - Reuse `set_mla_kv_buffer_triton(kv_buffer, loc, nope_part, rope_part)` for FP8 writes.

  3. **Cleanup:**
     - Remove the previously proposed FP8 scatter kernel and its env flag `SGLANG_USE_TRITON_FP8_SCATTER`.
     - Remove legacy tests and debug scripts for the old scatter path.

  ## Testing

  - **Byte-level equivalence**: `quantize_k_cache_separate` output matches the old concat+quantize path.
  - **End-to-end KV write** matches baseline behavior.
  - `pre-commit` passes.

 **Server:**
  ```bash
  python3 -m sglang.launch_server \
    --model-path deepseek-ai/DeepSeek-V3.2 \
    --trust-remote-code \
    --tp-size 8 \
    --dp-size 8 \
    --enable-dp-attention \
    --tool-call-parser deepseekv32 \
    --reasoning-parser deepseek-v3

  Benchmark:
  python3 -m sglang.bench_serving \
    --dataset-name random \
    --random-input-len 1024 \
    --random-output-len 1024 \
    --random-range-ratio 0.98 \
    --num-prompts 80 \
    --max-concurrency 16
```
  ## Results

  | Metric             | Baseline (disabled) | Optimized (default) | Improvement |
  |--------------------|---------------------|---------------------|-------------|
  | Benchmark duration | 153.09s             | 150.03s             | -2.0%       |
  | Total throughput   | 1059.16 tok/s       | 1080.93 tok/s       | +2.0%       |

  ## GPU Kernel Change (from nsys profile):
  - Removed: at::native::index_elementwise_kernel<...index_put_kernel_impl...>
  - Added: set_mla_kv_buffer_kernel (Triton)

  ## Correctness Validation:
  python3 benchmark/gsm8k/bench_sglang.py \
      --num-shots 8 \
      --num-questions 1319 \
      --parallel 1319
  - Accuracy: 0.951 (unchanged)
  - Invalid: 0.000

  ## Backward Compatibility

  - Default enabled: Users get performance improvement automatically
  - Fallback available: Set SGLANG_USE_TRITON_FP8_SCATTER=0 to disable
  - No API changes: Transparent to end users
  - Scope limited: Only affects MLA + FP8 KV cache path

  ### Related Issues

  Fixes #15104
  Related: #15025
   @Fridge003 
  ```

harvenstar committed:
Optimize FP8 MLA KV cache writes with Triton scatter kernel

This PR addresses the index_put_ overhead identified in #15104 by
replacing PyTorch index_put_ with a custom Triton scatter kernel.

Key improvements:
- Replaces ~1281 high-overhead index_put_ calls per decode step
- Reduces CPU dispatch overhead from ~40ms to negligible
- Performance: +1.9% throughput improvement (DeepSeek-V3.2)
- Default: Enabled with automatic fallback on errors
- Correctness: GSM8K accuracy unchanged (0.951)

Implementation details:
- Added set_mla_kv_buffer_fp8_triton() with explicit 2D stride handling
- Integrated Triton scatter in FP8 NSA KV write path
- Added try-except fallback to index_put_ on any error
- One-time logging for activation and fallback events

Addresses all reviewer feedback from draft:
- Use explicit error handling (if/raise) instead of assertions
- Clean up redundant logging imports
- Register SGLANG_USE_TRITON_FP8_SCATTER in Envs class
- Add comprehensive unit tests

Testing:
- Unit tests: test/manual/layers/attention/nsa/test_fp8_triton_scatter.py
- Benchmark: +1.9% throughput (1059.66 -> 1080.04 tok/s)
- Correctness: GSM8K acc=0.951 (unchanged)

Fixes #15104
Related: #15025

harvenstar committed:
Fix lint: apply black formatting

harvenstar committed:
Fix lint: remove unused imports

harvenstar committed:
Refactor FP8 MLA KV cache: avoid concat, reuse existing kernel

- Add quantize_k_cache_separate() to quantize k_nope/k_rope separately
- Reuse set_mla_kv_buffer_triton for FP8 writes (no new kernel needed)
- Remove set_mla_kv_buffer_fp8_triton and SGLANG_USE_TRITON_FP8_SCATTER
- Delete obsolete test files and debug scripts
- Net -98 lines: simpler, cleaner, same performance

Fridge003 commented on python/sglang/srt/layers/attention/nsa/quant_k_cache.py:
```diff
@@ -13,6 +13,113 @@ def quantize_k_cache(cache_k):
         return _quantize_k_cache_slow(cache_k)
 
 
+def quantize_k_cache_separate(
```
This implementation is not correct.
The quantization of k_nope and k_rope should be done in a single triton kernel.

You can refer to the implementation of `_quantize_k_cache_fast` kernel
The separate one just needs some simple refactors on the basis of this kernel

harvenstar replied:
@Fridge003 Thanks for the feedback. I've refactored the implementation:

  - Now uses a single Triton kernel by reusing the existing _quantize_k_cache_fast_kernel
  - Added _quantize_k_cache_fast_separate() that directly allocates two separate uint8 buffers
  - Avoids Python-side pack/view/split operations
  - Replaced all input validation asserts with explicit ValueError checks
  - Auto-fixes non-contiguous tensors

harvenstar replied:
The new implementation passes typed views to the existing kernel to avoid code duplication. Sanity tests pass with byte-wise equality vs the concat path.

harvenstar committed:
Refactor quantize_k_cache_separate to use single Triton kernel

Address reviewer feedback:
- Implement _quantize_k_cache_fast_separate() that directly outputs two
  separate tensors (nope_part, rope_part) in a single kernel call
- Reuse existing _quantize_k_cache_fast_kernel to avoid code duplication
- Replace all input validation asserts with explicit ValueError checks
- Auto-fix non-contiguous tensors with .contiguous()
- Add alignment checks for typed views (FP32/BF16)
- Delete test/manual/.../test_quantize_separate.py
- Move tests to __main__ section in quant_k_cache.py with byte-wise verification

The new implementation avoids Python-side pack/view/split operations
by allocating separate uint8 buffers and creating typed views for the
kernel to write into directly.

Tested: Sanity tests pass with byte-wise equality vs concat path.

harvenstar committed:
Fix lint: apply black formatting

harvenstar commented:
## Updated Performance Results 
@Fridge003 
  I've run the latest benchmark and the results show **significant improvement**:

  ### Performance Metrics

  | Metric | Baseline (disabled) | Optimized (current) | Improvement |
  |--------|---------------------|---------------------|-------------|
  | Benchmark duration | 153.35s | **146.84s** | **-4.25%**  |
  | Total throughput | 1059.16 tok/s | **1104.20 tok/s** | **+4.25%**  |


  **This is a 2x improvement compared to the previous result **

  ### GPU Kernel Changes (from nsys profile)

  **Removed:**
  - `at::native::index_elementwise_kernel<...index_put_kernel_impl...>` (PyTorch native indexing)
  - `CatArrayBatchedCopy` kernel

  **Added:**
  - `set_mla_kv_buffer_kernel` (Triton) - our optimized single-kernel implementation

  The kernel consolidation successfully eliminated the overhead from separate indexing and concatenation operations, resulting in cleaner execution and better performance and the profiler results support this. 
<img width="1620" height="573" alt="image" src="https://github.com/user-attachments/assets/0937ca0f-2c74-4bc9-9c58-a1d2aa4c3f33" />

Fridge003 commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -22,7 +22,10 @@
 from sglang.srt.layers.attention.nsa import index_buf_accessor
-from sglang.srt.layers.attention.nsa.quant_k_cache import quantize_k_cache
+from sglang.srt.layers.attention.nsa.quant_k_cache import (
+    quantize_k_cache,
```
Not need to import line26

harvenstar committed:
Fix kernel boundary bug in set_mla_kv_buffer_kernel for FP8 quantization

Problem:
The original two-branch logic assumed nope_dim % BLOCK == 0, which works
for non-FP8 (nope_dim=512, BLOCK=128). However, FP8 has nope_dim=528,
causing blocks to span the nope/rope boundary.

When pid_blk=4 (base=512, offs=[512,639]), the else branch would:
- Read entire block from rope (incorrect for offs [512,527])
- Create negative indices: offs_rope=[-16,...,111]
- Result: corrupted KV cache, accuracy drops to 0

Solution:
Use three-way branching:
1. Fast path: entire block in nope region (original logic preserved)
2. Fast path: entire block in rope region (original logic preserved)
3. Boundary case: per-element masking for cross-boundary blocks

This preserves performance for non-FP8 while fixing FP8 correctness.

harvenstar commented:
## Bug Fix: Kernel Boundary Handling for FP8 Quantization

  ### Issue Found
  After applying the optimization, accuracy dropped from 0.8 to 0 in evaluation tests.

  ### Root Cause
  The `set_mla_kv_buffer_kernel` had a boundary bug when handling FP8 quantization:
  - The original two-branch logic assumed `nope_dim % BLOCK == 0`
  - This works for non-FP8 (nope_dim=512, BLOCK=128) 
  - FP8 has nope_dim=528, causing blocks to span nope/rope boundary 
  - When pid_blk=4 (base=512), the kernel would read wrong data with negative offsets

  ### Fix Applied
  Changed to three-way branching:
  1. Fast path: entire block in nope region (preserved)
  2. Fast path: entire block in rope region (preserved)  
  3. Boundary case: per-element masking for cross-boundary blocks

  ### Result
  -  Accuracy restored to baseline (~0.85)

harvenstar committed:
Clean up redundant validation logic

Remove unnecessary checks that add no value:
- Hardcoded constant validations (nope_part_bytes == 528, etc.)
- Duplicate dimension checks already done in outer function
- Device/dtype/ndim validations (PyTorch handles these)
- Alignment checks for constants (512 % 4, 128 % 2 always true)
- Grid size validation (debug code)

Simplify verbose patterns:
- Squeeze logic: 12 lines → 2 lines
- Contiguous checks: use direct chaining

Result: -90 lines of redundant code, cleaner for review.

Fridge003 commented:
/tag-and-rerun-ci

harvenstar committed:
Merge branch 'main' into optimize/nsa-topk-paged-small-kernels
